### PR TITLE
ODrive

### PR DESCRIPTION
**Pull request by PhileasL**
# Purpose

This PR deals with the implementation of [ODrive](https://odriverobotics.com/) into drive node.

## What as been done ?

- Modification of drive node (from rad/s to turns/s),
- Implementation of a custom odrive library,
- odrive_ros sub-module deletion.

### Custom odrive library

This library is based on the work of [Framework-Robotics-GmbH](https://github.com/Framework-Robotics-GmbH). They come with a [ros2 node](https://github.com/Framework-Robotics-GmbH/odrive-ros2-node) that can communicates really easily with ODrive hardware using a serial communication. We [refactored this node](https://github.com/robotique-ecam/odrive-ros2-node) for it to stick to our architecture standards. By putting this node in our code base, we realized that it wasn't really a thing for a simple serial communication to have its own dedicated node.

So, we decided to transform this node into a [library](https://github.com/robotique-ecam/cdfr/blob/ODrive/src/modules/drive/include/ODrive.hpp) and use it directly from drive node.

As a simple serial communication through USB, we're expecting it to work pretty well. But in reality it's not the case. We were having some trouble when asking to ODrive to send a checksum with its message, 75% of messages returned will not have this checksum. We'll further investigate the problem in the future but for the moment, we're considering a simple check on the message length in order to validate it.

### Drive node modifications

As ODrive returning turns/s values, we've just modified all the formulas by switching from rad/s to turns/s. Navigation through simulation isn't broken, if simulation mode is enabled, we haven't forgotten to adapt the data into what simulation requires as input or returns.

We also add a gearbox ratio into formulas and the capability to change left motor as right motor and vice-versa without having to modify the wiring.

### YAML new parameters

- `odrive_usb_port` --> string --> the usb odrive port, ex: `odrive_usb_port: "/dev/ttyS1"`
- `invert_wheel` --> boolean --> the capability of changing right motor to left and vice versa, ex: `invert_wheel: false`
- `gearbox_ratio` --> float --> the gearbox ratio between motors and wheels, ex `gearbox_ratio: 8.75`

#### Note

I'll write a README explaining formulas and odometry readjustment after the cdfr.

### CODE REVIEW

File: src/modules/localisation/localisation/vlx/sensors_sim.py [hunk 1]
```diff
@@ -28,28 +28,31 @@ def __init__(self, node, addrs=[]):
...
             distances.update({addr: round(vlx.getValue(), 4)})
         return distances
 
     def get_time_stamp(self):
-        """ Get time stamp from Webots """
+        """Get time stamp from Webots"""
         self.node.robot.step(0)
         t = self.node.robot.getTime()
-        return Time(sec=int(t), nanosec=int((t % int(t)) * 1e9))
+        if int(t) != 0:
```
Comment (3wnbr1) [inline src/modules/localisation/localisation/vlx/sensors_sim.py:42]: No if statement, refactor to use `int((t - int(t)) * 1e9)` instead which doesn't fault for `t = 0`

File: src/modules/drive/src/drive.cpp [hunk 2]
```diff
@@ -74,44 +77,34 @@ Drive::Drive() : Node("drive_node") {
...
-  this->declare_parameter("steps_per_turn");
-  this->declare_parameter("microsteps");
   this->declare_parameter("wheels.separation");
   this->declare_parameter("wheels.radius");
-  this->declare_parameter("microcontroler.max_steps_frequency");
-  this->declare_parameter("microcontroler.speedramp_resolution");
 
 // Get parameters from yaml
 #ifndef SIMULATION
   this->declare_parameter("i2c_bus");
```
Comment (3wnbr1) [inline src/modules/drive/src/drive.cpp:88]: Remove I2C if not used

File: src/modules/drive/src/drive.cpp [hunk 3]
```diff
@@ -1,7 +1,5 @@
 #include <drive.hpp>
 
-#define I2C_ADDR_MOTOR_LEFT 0x10
-#define I2C_ADDR_MOTOR_RIGHT 0x11
 #define I2C_ADDR_SLIDER 0x12
```
Comment (3wnbr1) [inline src/modules/drive/src/drive.cpp:3]: Sliders / actuators shall be moved to a separate node now that there is no more I2C required by drive. Was created like this to have only one process accessing the file descriptor `/dev/i2c1`

File: src/modules/drive/src/ODrive.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        return 0;
+    }
+    RCLCPP_INFO(node->get_logger(), "could not set velocity: %f", velocity);
+    return -1;
+}
+
+/*
+return tuple of position and velocity, returns both as -1 if request fails
+*/
+std::pair<float, float> ODrive::getPosition_Velocity(const int motor)
```
Comment (3wnbr1) [inline src/modules/drive/src/ODrive.cpp:213]: NIT : Method naming convention

File: src/modules/drive/include/drive.hpp [hunk 5]
```diff
@@ -2,13 +2,19 @@
 #define DRIVE_NODE_HPP
 
 #ifndef SIMULATION
+
+#include <ODrive.hpp>
 #include "i2c.hpp"
```
Comment (3wnbr1) [inline src/modules/drive/include/drive.hpp:7]: See comment below about integrating Sliders / actuators into a separate node

**Review by 3wnbr1 — CHANGES_REQUESTED** [review]
Good implementation, few changes requested and would be much better to move the sliders code into actuators node

File: src/modules/localisation/localisation/vlx/sensors_sim.py (hunk 1 above)
Comment (PhileasL) [reply src/modules/localisation/localisation/vlx/sensors_sim.py:42]: Yep, haven't thought about it

**Comment by PhileasL** [comment]
As it's not concerning ODrive, what about dedicating another PR for the separation sliders / actuators from drive  ?

**Comment by 3wnbr1** [comment]
> As it's not concerning ODrive, what about dedicating another PR for the separation sliders / actuators from drive ?

Agreed, you can strip either strip it now or in the dedicated PR

**Review by 3wnbr1 — APPROVED** [review]

**Comment by PhileasL** [comment]
> Agreed, you can strip either strip it now or in the dedicated PR

Yep, it'll have its dedicated PR which will also contains a bunch of overall modifications for physical robots to work.

**Pull request merged**